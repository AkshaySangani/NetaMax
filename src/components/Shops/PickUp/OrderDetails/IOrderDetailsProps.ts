import { IPickedUpOrderDetail } from "dataflows/Shops/DeliverOrder/IPickedUpOrderDetail";

export interface IOrderDetailsProps {
  /**
   * Redirect to the next step
   * @type {() => void}
   **/
  onVerify: (number: number) => void;

  /**
   * The picked up order detail
   * @type {IPickedUpOrderDetail}
   **/
  orderDetail: IPickedUpOrderDetail | undefined;
}
