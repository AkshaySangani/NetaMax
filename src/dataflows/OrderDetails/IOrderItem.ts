export interface IOrderItem {
  /**
   * The Item's name
   * @type {String}
   *
   * */
  name: string;
  /**
   * The Item's id
   * @type {String}
   *
   * */
  product_id?: number;
  /**
   * The Item's quantity bought
   * @type {String}
   *
   * */
  quantity: number;
  /**
   * The Item's image
   * @type {String}
   *
   * */
  image?: string;
  /**
   * The Items's unit price including taxes
   * @type {Number}
   *
   * */
  unit_price_incl_tax: number;
}
