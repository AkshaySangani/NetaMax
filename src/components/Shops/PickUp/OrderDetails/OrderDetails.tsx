import {
  DrawerHeader,
  Flex,
  DrawerBody,
  Container,
  Box,
  VStack,
  DrawerFooter,
  Text,
} from "@chakra-ui/react";
import { DeliverButton } from "components/Shops/DeliverOrderButton/DeliverOrderButton.styled";
import { OrderDetail } from "components/Shops/OrderDetail/OrderDetail";
import { CURRENCY_NAME } from "constants/localeFormatConstants";
import { PickUpWizardSteps } from "constants/pickUpTokenConstants";
import { ReactElement } from "react";
import IconPickUpBasket from "styled/icons/Shops/PickUp/Basket";
import { formatMoney } from "utils/currencyUtils";
import { IOrderDetailsProps } from "./IOrderDetailsProps";

/**
 * The OrderDetails Component
 * @param {IOrderDetailsProps} orderDetailsProps The orderDetails props.
 * @returns {ReactElement} The OrderDetails component
 */
export const OrderDetails = (orderDetailsProps: IOrderDetailsProps): ReactElement => {
  const { orderDetail, onVerify } = orderDetailsProps;
  return (
    <>
      <DrawerHeader paddingX="16px">
        <Flex alignItems="center">
          <Text fontSize="16px" data-testid="order-details-header">
            Productos a entregar
          </Text>
        </Flex>
      </DrawerHeader>
      <DrawerBody paddingX="0px">
        <Container paddingX="16px">
          <Text fontSize="14px" fontWeight="400" color="#353535">
            Cliente
          </Text>
          <Text
            fontWeight="800"
            fontSize="24px"
            lineHeight="30px"
            data-testid="order-details-client"
          >
            {orderDetail?.customer?.firstName}
          </Text>
          <Flex marginTop="8px">
            <Box mr="5px" height="18px">
              <IconPickUpBasket />
            </Box>
            <Text fontWeight="500" fontSize="14px" lineHeight="18px">
              {orderDetail?.productsAmount} Productos a entregar
            </Text>
          </Flex>
        </Container>
        <VStack
          marginTop="16px"
          paddingTop="16px"
          borderY="2px solid rgba(53, 53, 53, 0.1)"
          height="50%"
          overflowY="auto"
        >
          {orderDetail?.productsOrder.map((order, index) => (
            <OrderDetail key={`order-${index}`} {...order} />
          ))}
        </VStack>
        <Container marginTop="18px" paddingX="16px">
          <Text fontSize="14px" fontWeight="400" color="#353535">
            Informaci√≥n de la orden
          </Text>
          <Flex marginTop="9px" alignItems="center" justify="space-between">
            <Text fontSize="14px" fontWeight="600">
              Subtotal
            </Text>
            <Text fontSize="16px" fontWeight="800">
              {formatMoney(Number(orderDetail?.orderSubtotalInclTax), CURRENCY_NAME)}
            </Text>
          </Flex>
          <Flex
            borderBottom="0.5px solid rgba(53, 53, 53, 0.2)"
            alignItems="center"
            justify="space-between"
            paddingBottom="8px"
          >
            <Text fontSize="14px" fontWeight="600">
              Descuento
            </Text>
            <Text fontSize="16px" fontWeight="800">
              {formatMoney(Number(orderDetail?.orderDiscount), CURRENCY_NAME)}
            </Text>
          </Flex>
          <Flex marginTop="9px" alignItems="center" justify="space-between">
            <Text fontSize="14px" fontWeight="600">
              Total
            </Text>
            <Text fontSize="16px" fontWeight="800">
              {formatMoney(Number(orderDetail?.orderTotal), CURRENCY_NAME)}
            </Text>
          </Flex>
        </Container>
      </DrawerBody>
      <DrawerFooter>
        <Box width="100%">
          <DeliverButton
            onClick={() => onVerify(PickUpWizardSteps.ORDER_CONFIRMED)}
            height="48px"
            display="flex"
            alignItems="center"
            cursor="pointer"
          >
            <Text fontWeight="600">Confirmar entrega</Text>
          </DeliverButton>
        </Box>
      </DrawerFooter>
    </>
  );
};
