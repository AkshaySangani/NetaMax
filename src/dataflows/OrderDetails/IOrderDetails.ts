import { IOrderItem } from "./IOrderItem";

export interface IOrderDetails {
  /**
   * The order items array
   * @type {IOrderItem[]}
   */
  items: IOrderItem[];

  /**
   * The order's total
   * @type {number}
   */
  order_total: number;

  /**
   * The order's subtotal excluding tax
   * @type {number}
   */
  order_subtotal_excl_tax: number;

  /**
   * The order's discount
   * @type {number}
   */
  order_discount: number;
}
