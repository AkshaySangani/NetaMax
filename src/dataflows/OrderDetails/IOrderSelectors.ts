import { RootState } from "state/store";

import { IOrderDetails } from "./IOrderDetails";

/**
 * Selector to get the order items from the state.
 * @param {RootState} state the root state.
 * @returns {any} the selected orders.
 */
export const selectOrder = (state: RootState): IOrderDetails | undefined =>
  state?.nopOrders?.orderDetails;

/**
 * Selector to get the isLoadingOrder state from the state.
 * @param {RootState} state the root state
 * @returns {boolean} the isLoadingOrder state
 */
export const selectIsLoadingOrder = (state: RootState): boolean => state?.nopOrders?.isLoadingOrder;
