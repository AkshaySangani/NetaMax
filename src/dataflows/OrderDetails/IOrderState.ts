import { IOrderDetails } from "./IOrderDetails";

export interface IOrderState {
  /**
   * The order details object
   * @type {IOrderDetails[]}
   */
  orderDetails: IOrderDetails;

  /**
   * indicates if is loading the orders call
   * @type {boolean}
   **/
  isLoadingOrder: boolean;
}
