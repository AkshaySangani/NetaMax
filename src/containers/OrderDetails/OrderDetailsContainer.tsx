import React, { useEffect, ReactElement } from "react";

import { useDispatch, useSelector } from "react-redux";

import { BackToHome } from "components/common/BackHomeComponent/BackToHome";
import { Loader } from "components/common/Loader";
import { OrderTable } from "components/OrderDetails/OrderTable/OrderTable";
import { HeaderContainer } from "components/ProductDetails/ProductDetails.styled";
import { selectIsLoadingToken } from "dataflows/NopAuth/NopAuthSelectors";
import { acquireNopToken } from "dataflows/NopAuth/NopAuthThunks";
import { selectOrder } from "dataflows/OrderDetails/IOrderSelectors";
import { getOrderItems } from "dataflows/OrderDetails/OrderThunks";

import { Img } from "@chakra-ui/image";
import { Box, Center, Container, Text } from "@chakra-ui/layout";

import { IOrderDetailsContainerProps } from "./IOrderDetailsContainer";

/**
 * The order details container.
 * @param {IOrderDetailsContainerProps} props the component props.
 * @returns { ReactElement } The order details container.
 */
export const OrderDetailsContainer = (props: IOrderDetailsContainerProps): ReactElement => {
  const { orderId } = props;
  const dispatch = useDispatch();
  const orderItems = useSelector(selectOrder);
  const isLoadingToken = useSelector(selectIsLoadingToken);

  /**
   * Acquire NopCommerce token to handle requests.
   * @returns {void}
   */
  const authenticateNopUser = (): void => {
    //TODO: The following credentials should be encrypted,
    //for urgency reasons they are added in code temporarily.
    dispatch(
      acquireNopToken({
        email: "admin@neta.mx",
        password: "2022.SerieA.Net@",
      })
    );
  };

  /**
   * Get Order Details
   * @param {string} orderId the order id
   * @returns {void}
   */
  const getOrderDetails = (orderId: string): void => {
    dispatch(getOrderItems(orderId));
  };

  useEffect(() => {
    authenticateNopUser();
  }, []);

  useEffect(() => {
    isLoadingToken === false && getOrderDetails(orderId);
  }, [isLoadingToken]);

  return (
    <div>
      {isLoadingToken ? (
        <Center h="100%" w="100%" position="fixed">
          <Box>
            <Box pb="25px">
              <Img src="/assets/images/logo.png" width="120px" />
            </Box>
            <Center>
              <Loader />
            </Center>
          </Box>
        </Center>
      ) : (
        <Container centerContent={false} maxW="container.lg" p="6">
          <HeaderContainer>
            <BackToHome size={18} weight={700} />
          </HeaderContainer>
          {/* <SplitCard {...splitCardObject} /> */}
          <Text fontSize="20px" fontWeight="700">
            N. de orden: {orderId}
          </Text>
          {orderItems && (
            <OrderTable
              items={orderItems.items}
              order_total={orderItems.order_total}
              order_discount={orderItems.order_discount}
              order_subtotal_excl_tax={orderItems.order_subtotal_excl_tax}
            />
          )}
        </Container>
      )}
    </div>
  );
};
