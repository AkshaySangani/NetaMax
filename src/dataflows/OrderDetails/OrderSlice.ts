import { createSlice, PayloadAction } from "@reduxjs/toolkit";

import { IOrderDetails } from "./IOrderDetails";
import { IOrderState } from "./IOrderState";
import { getOrderItems } from "./OrderThunks";

const initialState: IOrderState = {
  orderDetails: {
    items: [],
    order_total: 0,
    order_subtotal_excl_tax: 0,
    order_discount: 0,
  },
  isLoadingOrder: false,
};

const nopOrdersSlice = createSlice({
  name: "nopOrders",
  initialState,
  reducers: {
    selectOrder: (state, action: PayloadAction<IOrderDetails>) => {
      state.orderDetails = action.payload;
    },
  },
  extraReducers: {
    [getOrderItems.pending.type]: (state) => {
      state.isLoadingOrder = true;
    },
    [getOrderItems.fulfilled.type]: (state, action) => {
      state.orderDetails = action.payload;
      state.isLoadingOrder = false;
    },
  },
});

/**
 * Actions
 */
export const { selectOrder } = nopOrdersSlice.actions;

/**
 * Reducers
 */
export const nopOrdersReducer = nopOrdersSlice.reducer;
