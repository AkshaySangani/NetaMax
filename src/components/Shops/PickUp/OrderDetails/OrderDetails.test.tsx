import { Drawer } from "@chakra-ui/react";
import { render, cleanup, screen } from "@testing-library/react";
import { OrderDetails } from "./OrderDetails";
import { IOrderDetailsProps } from "./IOrderDetailsProps";

describe("OrderDetails component", () => {
  let mock: IOrderDetailsProps;
  const isOpen = true;
  const onClose = jest.fn();
  const onVerify = jest.fn();

  // eslint-disable-next-line require-jsdoc
  const renderOrderDetails = () => {
    return (
      <Drawer
        isOpen={isOpen}
        placement="right"
        onClose={onClose}
        size="md"
        closeOnEsc
        closeOnOverlayClick
      >
        <OrderDetails {...mock} />
      </Drawer>
    );
  };

  beforeEach(() => {
    mock = {
      onVerify,
      orderDetail: {
        name: "",
        orderTotal: 0,
        orderSubtotalInclTax: 0,
        orderDiscount: 0,
        productsAmount: 0,
        productsOrder: [],
        customer: {
          id: "1063507",
          name: "asd",
          isNewUser: true,
          ZipPostalCode: "asd",
          username: "541134375450",
          email: "541134375450",
          billingAddress_Id: "86287",
          shippingAddress_Id: "",
          customerGuid: "027ce196-3b7c-409a-b33f-0e2966cf8c5f",
          isTaxExempt: true,
          affiliateId: "",
          vendorId: "",
          hasShoppingCartItems: false,
          deleted: false,
          registeredInStoreId: "",
          customerCurrentNetaWallet: "0.000000000000000",
          referralCode: "Milo-Molina-7620",
          firstName: "Milo Molina",
        },
      },
    };
  });

  afterEach(() => {
    cleanup();
  });

  test("OrderDetails should be render correctly", () => {
    const { container } = render(renderOrderDetails());
    expect(container).toMatchSnapshot();
  });

  test("Header should be render properly", () => {
    render(renderOrderDetails());
    expect(screen.getByTestId("order-details-header")).toHaveTextContent(/Productos a entregar/);
  });

  test("Body should be render properly", () => {
    render(renderOrderDetails());
    expect(screen.getByText("InformaciÃ³n de la orden")).toBeInTheDocument();
    expect(screen.getByText("Cliente")).toBeInTheDocument();
    expect(screen.getByText("Subtotal")).toBeInTheDocument();
    expect(screen.getByText("Descuento")).toBeInTheDocument();
    expect(screen.getByText("Total")).toBeInTheDocument();
  });

  test("Footer should be render properly", () => {
    render(renderOrderDetails());
    expect(screen.getByText("Confirmar entrega")).toBeInTheDocument();
  });
});
