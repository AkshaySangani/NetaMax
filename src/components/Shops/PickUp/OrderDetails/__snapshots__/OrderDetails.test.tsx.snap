// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderDetails component OrderDetails should be render correctly 1`] = `<div />`;
